### ISSUE TITLE
Something is critically wrong with chemical reactions.

### ISSUE THREAD

**Issue by Winterous** [tgstation/tgstation#39045]
*edit* I thought it might be helpful if I provided a link to the file. All I've done is add in admin logging procs in various places, which you can find easily by searching "admin".
https://github.com/Winterous/tgstation/blob/Winterous-patch-6/code/modules/reagents/chemistry/holder.dm

So I've been going through the holder.dm code, adding in debug logging to try and figure out why #38216 happens, and...
I've cut this down to save space because it's quite large, but essentially; 100 unit container, add 10, 10, 10 chlorine, 10,10,10 phosphorus, 10,10 of radium (making mutagen), and then as follows.

ADMIN LOG: adding 10 units of radium
ADMIN LOG: total volume now 90
ADMIN LOG: removing 10 units of radium from 10
ADMIN LOG: post clamp, removing 10 units of radium from 10
ADMIN LOG: total volume now 80
ADMIN LOG: removing 10 units of phosphorus from 10
ADMIN LOG: post clamp, removing 10 units of phosphorus from 10
**ADMIN LOG: total volume now 130**
ADMIN LOG: removing 10 units of chlorine from 10
ADMIN LOG: post clamp, removing 10 units of chlorine from 10
ADMIN LOG: total volume now 120
ADMIN LOG: existing 120 + new 30 of mutagen > max volume 100
ADMIN LOG: reducing amount of mutagen to -20
ADMIN LOG: lol wat amount of mutagen is now -20
 The solution begins to bubble.
(essentially the result is that the third batch of 30 mutagen vanishes, leaving you with 60 instead of 90)

Of significant note: doing 10,10 chlorine, 10,10 phosphorus, 10,10 radium, and THEN 10/10/10 gives you 90 units correctly; my thought was that something about changing the amount of a reagent, but not completely removing it, messes with how total volume is calculated?

The next thing I was gonna look into was update_total I think, but already when the map boots up it runs through like 2,500 admin logs of reagent containers spawning in, so doing it for updates would require something less demanding, or a controlled environment (which I don't know how to do).

All the debug logging is on my patch 6 branch, but after doing this investigation I'm still not able to figure out what the issue is. Someone who knows the code please get on this, it's a super annoying issue, and hopefully this has narrowed down the problem a lot.

**Comment by Hate9** [comment]
Holy hell, what the shit is going on?

**Comment by Tlaltecuhtli** [comment]
were you using macros?

**Comment by Winterous** [comment]
No, macros are exclusively used by homos.

**Comment by Winterous** [comment]
SO I MANAGED TO SNEAK A PEEK AT EXACTLY WHAT IS GOING WRONG HERE, WITH EXCESSIVE DEBUG LOGGING!

Annie Holmes says, "Blap"
ADMIN LOG: updating_totalx2
ADMIN LOG: adding 10 units of radium
ADMIN LOG: updating_total
ADMIN LOG: deleting Chlorine
ADMIN LOG: Chlorine volume 10, total now 10
ADMIN LOG: deleting Phosphorus
ADMIN LOG: Phosphorus volume 10, total now 20
ADMIN LOG: deleting Unstable mutagen
ADMIN LOG: Unstable mutagen volume 60, total now 80
ADMIN LOG: deleting Radium
ADMIN LOG: Radium volume 10, total now 90
ADMIN LOG: total volume now 90
ADMIN LOG: removing 10 units of radium from 10
ADMIN LOG: post clamp, removing 10 units of radium from 10
ADMIN LOG: updating_total
ADMIN LOG: deleting Chlorine
ADMIN LOG: Chlorine volume 10, total now 10
ADMIN LOG: deleting Phosphorus
ADMIN LOG: Phosphorus volume 10, total now 20
ADMIN LOG: deleting Unstable mutagen
ADMIN LOG: Unstable mutagen volume 60, total now 80
ADMIN LOG: deleting Radium
ADMIN LOG: updating_total
ADMIN LOG: deleting Chlorine
ADMIN LOG: Chlorine volume 10, total now 10
ADMIN LOG: deleting Phosphorus
ADMIN LOG: Phosphorus volume 10, total now 20
ADMIN LOG: deleting Unstable mutagen
ADMIN LOG: Unstable mutagen volume 60, total now 80
ADMIN LOG: total volume now 80
ADMIN LOG: removing 10 units of phosphorus from 10
ADMIN LOG: post clamp, removing 10 units of phosphorus from 10
ADMIN LOG: updating_total
ADMIN LOG: deleting Chlorine
ADMIN LOG: Chlorine volume 10, total now 10
ADMIN LOG: deleting Phosphorus
**ADMIN LOG: updating_total
ADMIN LOG: deleting Chlorine
ADMIN LOG: Chlorine volume 10, total now 10
ADMIN LOG: deleting Unstable mutagen
ADMIN LOG: Unstable mutagen volume 60, total now 70
ADMIN LOG: deleting Unstable mutagen
ADMIN LOG: Unstable mutagen volume 60, total now 130
ADMIN LOG: total volume now 130**
ADMIN LOG: removing 10 units of chlorine from 10
ADMIN LOG: post clamp, removing 10 units of chlorine from 10
ADMIN LOG: updating_total
ADMIN LOG: deleting Chlorine
**ADMIN LOG: updating_total
ADMIN LOG: deleting Unstable mutagen
ADMIN LOG: Unstable mutagen volume 60, total now 60
ADMIN LOG: deleting Unstable mutagen
ADMIN LOG: Unstable mutagen volume 60, total now 120
ADMIN LOG: total volume now 120
ADMIN LOG: existing 120 + new 30 of mutagen > max volume 100
ADMIN LOG: reducing amount of mutagen to -20
ADMIN LOG: lol wat amount of mutagen is now -20**
 The solution begins to bubble.
ADMIN LOG: updating_total
ADMIN LOG: deleting Unstable mutagen
ADMIN LOG: Unstable mutagen volume 60, total now 60

So basically every single time something happens, it runs through the update_total proc, which zeroes out the contents of a container and readds it all reagent by reagent. For some reason it likes to run through that proc twice in a row for no apparent reason, but whatever.

What's happening is in bold; for some reason, when you add the third lot of radium, the update_total proc decides to count the mutagen towards the total volume of the container TWICE.

**Comment by Winterous** [comment]
And keep in mind this doesn't always happen, depending upon exactly when you put in reagents, it can decide to do this, or not do this. I think this is completely beyond my ability to understand, but hopefully now that I've gone to the effort of identifying where things go wrong, someone can finally fix this issue.

**Comment by vuonojenmustaturska** [comment]
i should rewrite reactions

**Comment by bgobandit** [comment]
@Iamgoofball

**Comment by Iamgoofball** [comment]
what the absolute fuck

**Comment by Hate9** [comment]
At the very least, SOMEONE should rewrite reactions. This is terrible.

**Comment by Winterous** [comment]
Presumably one of you guys is the expert, hopefully this is enough for you to figure it out.

**Comment by Supermichael777** [comment]
Technically it does this three times in mobs

**Comment by Supermichael777** [comment]
Holy fuck how did it end up in the list twice?

**Comment by Supermichael777** [comment]
What were you adding reagents to, with what were you adding them, because something isn't using the holder procs

**Comment by Winterous** [comment]
I've got the process I used in the original post.

> I've cut this down to save space because it's quite large, but essentially; 100 unit container, add 10, 10, 10 chlorine, 10,10,10 phosphorus, 10,10 of radium (making mutagen), and then as follows.

It's something to do with adding more of a reagent to an existing reagent, because it doesn't happen it you do it in lots of 30 instead of 10.

**Comment by jegub** [comment]
Seems like the cause is the space that crept in at the start of line 554 of holder.dm.
